Finetuning

jamesrequa opened the pull request:
Using VGG16 pre-trained on imagenet for fine tuning on the Planet Amazon dataset. Achieved 0.928 fbeta score after just 10 epochs (using 128x128 images).

jamesrequa committed:
finetuning - densenet169

jamesrequa committed:
updated classifier

jamesrequa committed:
changed batch size

jamesrequa committed:
fix to map_predictions

jamesrequa committed:
added save model

jamesrequa committed:
normalize between 0 and 1

jamesrequa committed:
changes to epochs per training run

jamesrequa committed:
Updates

jamesrequa committed:
updated augmentations

- getting better results with `shear_range` instead of `zoom_range`

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -190,119 +191,121 @@
-    train_losses += tmp_train_losses
-    val_losses += tmp_val_losses
+callbacks = [EarlyStopping(monitor='val_loss', patience=2, verbose=1, min_delta=1e-4),
+             ReduceLROnPlateau(monitor='val_loss', factor=0.5, patience=1, cooldown=0, verbose=1),
```
If we use `ReduceLROnPlateau` we don't need the `for` loop over our array of epochs/learning_rate right? I guess we can get rid of this thing :D

godardt replied:
I see things such as `Epoch 00004: reducing learning rate to 4.999999873689376e-05.` So I believe we can definitely remove our for loop :D

jamesrequa replied:
I actually had originally thought it would still be nice to have a "learning rate schedule" on top of the `ReduceLROnPlateau`. Right now I'm retraining without the loop to see how `ReduceLROnPlateau` does on its own, but in this case I changed `factor` to 0.1 so that the learning rate decreases more (effectively creating our learning rate schedule on the fly).

jamesrequa replied:
In order for `ReduceLROnPlateau` to be compatible with `EarlyStopping` we can actually use the `min_lr` argument to set a lower bound on the learning rate (i.e. `min_lr=1e-7`) so that `ReduceLROnPlateau` will stop reducing the learning rate beyond that.

jamesrequa replied:
And actually they are technically compatible anyway because I have set `min_delta=1e-4` for `EarlyStopping` this means that unless `val_loss` improves by at least 1e-4 then its not considered any improvement meanwhile `ReduceLROnPlateau` will not reduce the lr as long as there is some small improvement regardless of how small it is.

godardt replied:
Oh I see. On my side I only removed it without trying to make them both compatible with each other. My bad :(

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -190,119 +191,121 @@
-# ## Load Best Weights
+for learn_rate, epochs in zip(learn_rates, epochs_arr):
+    model.compile(optimizer=Adam(lr=learn_rate), loss='binary_crossentropy', metrics = ['accuracy'])
+    model.fit_generator(train_generator, steps, epochs, verbose=1, 
```
I get an error here `Unable to open file: name = 'weights/weights.best.hdf5'`. I believe this is because you didn't commit the `weights` folder. In git if you try to commit an empty folder it will be ignored (I know it's kinda weird). You have to put 1 file in the folder at least. If you don't know what kind of "junk file" you want to put in the folder you can put either a `.DS_Store` or a `__init__.py` for folders containing python source files :)

jamesrequa replied:
Strange I actually did include the weights folder in the repo when I committed but I guess it didn't show up because the folder was empty like you say. So this time I'll make sure there is a file in there :)

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -190,119 +191,121 @@
-# <markdowncell>
-
-# ## Make predictions
+def predict(batch_size=128):
```
Why did you move this method out of the `keras_helper.py` file? 🤔

godardt replied:
Oh it's because you are applying it to the VGG model right? In this case see my comment [here](https://github.com/EKami/planet-amazon-deforestation/pull/14/files/9bee972a40beaa772e14b7595697ec163c1e49c9#r128918990), we should move this method to it's own file too :)

jamesrequa replied:
This is because I'm not using the `AmazonKerasClassifier` class. The same goes for `map_predictions`. Only these two functions are being used from `keras_helper` file so I just moved them into the notebook. For this reason we could probably do without the `keras_helper` file entirely...Or do you have a better idea?

jamesrequa replied:
For VGG model, its only like 8 lines of code (not counting the imports), do you think its really so much that it needs to be in a separate file and not just have it in the notebook? The reason DenseNet is its own separate file is because its very long and not a part of Keras so I couldn't just import it easily like I do with VGG.

godardt replied:
Sorry I didn't commit my message in time, please refer to my comment [here](https://github.com/EKami/planet-amazon-deforestation/pull/14#discussion_r128919042) :)

godardt replied:
Yes as you can also move the predict/map prediction methods to their own file :) . Creating a new file for a separate context is never a bad idea. Keeping multiples contexts into 1 file is :)

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -313,21 +316,31 @@
+# <codecell>
 
-# <markdowncell>
+def map_predictions(predictions, thresholds):
```
Same here

godardt replied:
See my comment [here](https://github.com/EKami/planet-amazon-deforestation/pull/14/files/9bee972a40beaa772e14b7595697ec163c1e49c9#r128918990), we should move this method to it's own file too :)

jamesrequa replied:
Yea this is because this function and `predict` are the only two functions I am using from `keras_helper`. I am not sure we should still keep that file just for two functions?

godardt replied:
Moving them to a separate file like `vgg16.py` could be a good idea. Moving everything related to the VGG model to its own file is great I think :)

jamesrequa replied:
OK I'll create a new file called `vgg16.py` which contains the model and also the two functions `predict` and `map_predictions`.

godardt commented on src/custom_layers.py:
```diff
@@ -0,0 +1,72 @@
+
+        out = in * gamma + beta,
+
+    where 'gamma' and 'beta' are the weights and biases larned.
```
Typo here on `larned` :)

godardt commented on src/custom_layers.py:
```diff
@@ -0,0 +1,72 @@
+    from keras import initializers as initializations
+import keras.backend as K
+
+class Scale(Layer):
```
Did you write this class by your own? If you didn't it's always a good idea to put the link where you found it in the document for later referencing :) . I'm a bit curious about why we need this class 🤔

jamesrequa replied:
This class was created to help us convert DenseNet from caffe model to Keras. I am actually not using it in this PR so I can remove it and the DenseNet model as well.

godardt commented on src/data_helper.py:
```diff
@@ -105,7 +104,14 @@ def get_train_generator(self, batch_size):
+                    # Zero-center by mean pixel
+                    img_array[:, :, 0] -= 103.939
+                    img_array[:, :, 1] -= 116.779
+                    img_array[:, :, 2] -= 123.68
```
Oh nice! :octocat:  Does it means that we'll use the mean for **all** models from now? (Vgg, Resnet...)

jamesrequa replied:
Yes this same preprocessing works for all imagenet pre-trained models which is basically every model we would want to use for finetuning :)

godardt replied:
Nice!! :D But I think didn't get the whole story around why some ppl decides to normalize the values, other decides to remove the mean other do something else blah blah. I'm still a bit confused even if I understand that here we are forced to do that for finetunning.

godardt commented on src/data_helper.py:
```diff
@@ -181,7 +194,7 @@ def _get_validation_split(self):
         labels = list(set(flatten([l.split(' ') for l in train['tags'].values])))
         label_map = {l: i for i, l in enumerate(labels)}
-        inv_label_map = {i: l for l, i in label_map.items()}
+        #inv_label_map = {i: l for l, i in label_map.items()}
```
We can remove this :)

godardt commented on src/data_helper.py:
```diff
@@ -144,10 +151,16 @@ def get_prediction_generator(self, batch_size):
+                    # Zero-center by mean pixel
+                    img_array[:, :, 0] -= 103.939
+                    img_array[:, :, 1] -= 116.779
+                    img_array[:, :, 2] -= 123.68
```
I really fucked it with this. You shouldn't have to modify the same code in 2 places :(

godardt commented on src/data_helper.py:
```diff
@@ -144,10 +151,16 @@ def get_prediction_generator(self, batch_size):
+                    img_array[:, :, 0] -= 103.939
+                    img_array[:, :, 1] -= 116.779
+                    img_array[:, :, 2] -= 123.68
+                    img_array = img_array / 255
```
Why now we divide by 255? 🤔

jamesrequa replied:
I have read that its still a good idea to have all image arrays as smaller numbers so this is why we divide by 255 after subtracting the mean.

godardt replied:
Oh ok. I still don't really get it lol. Why would smaller numbers would be quicker to process? I guess I'll figure this out with time ☹️

godardt commented on src/data_helper.py:
```diff
@@ -264,7 +285,14 @@ def _val_transform_to_matrices(self, *args):
+        # Zero-center by mean pixel
+        img_array[:, :, 0] -= 103.939
+        img_array[:, :, 1] -= 116.779
+        img_array[:, :, 2] -= 123.68
```
Uh.. 3 places. This code definitely needs a refactoring, sorry for that :(

godardt commented on src/data_helper.py:
```diff
@@ -264,7 +285,14 @@ def _val_transform_to_matrices(self, *args):
+        img_array[:, :, 0] -= 103.939
+        img_array[:, :, 1] -= 116.779
+        img_array[:, :, 2] -= 123.68
+        img_array = img_array / 255
```
Again, why dividing by 255 for the test/val set and not with the train set? 🤔

jamesrequa replied:
We are dividing by 255 in the training set if you check the `ImageDataGenerator` we use `rescale=1./255`. So in this way actually training/validation/predicitons all consistently divide by 255.

godardt commented on src/densenet169.py:
```diff
@@ -0,0 +1,214 @@
+    return concat_feat, nb_filter
+
+
+def create_model():
```
I don't see this method being called anywhere in the code. Are you sure we really need it? If we don't we should remove it as it will just add an "information junk burden" to the code :)

jamesrequa replied:
Yes we don't use it because in the branch I was using DenseNet model but in the PR I switched to just use VGG and no longer using DenseNet. I will remove DenseNet model file and custom layer files.

godardt reviewed (changes requested):
Very good PR!! Good job! 🎉

EKami commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -190,119 +191,121 @@
 
-plt.plot(train_losses, label='Training loss')
-plt.plot(val_losses, label='Validation loss')
-plt.legend();
```
The plot, why did they disappear? 😱 🙀 
We can still pass the `History` callback to the training right?

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -190,119 +191,121 @@
 
-plt.plot(train_losses, label='Training loss')
-plt.plot(val_losses, label='Validation loss')
-plt.legend();
```
The plot, why did they disappear? 😱 🙀
We can still pass the History callback to the training right?

jamesrequa replied:
I honestly didn't find much use for it and because I was dealing with limited memory allocation I didn't want to waste any memory where it wasn't absolutely essential for training (i.e. so we can use a bigger batch size). If you disagree then I can add the history callback back in including the plot.

godardt replied:
Oh, I see. Actually I can augment your docker limit from 15gb to something like 20gb (it'll use the swap space tho). If I do that could we put back the curves? They were very useful to me 😢 . Deal?

godardt replied:
Btw I don't think they take up too much space memory. It was only taking 2 `float32` per epochs (1 for the `train_loss`, the other for the `val_loss`). So lets say we ran it on 30 epochs = 30 * 4 = 120 bytes = 0.00012mb of memory space :D

jamesrequa replied:
OK no worries I am adding back the history callback in along with the plot :)

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -190,119 +191,121 @@
+learn_rates = [1e-4, 1e-5, 1e-6]
+batch_size = 128
+train_generator = preprocessor.get_train_generator(batch_size)
+steps = len(X_train) // batch_size
```
Using `//` isn't a good idea I think. I did that by the past (because intuitively it makes sense) and it resulted in Keras not training on the whole data. For whatever reason Keras figure out how to deal with `float` values for the steps. I know it's weird but it works 😛 .

jamesrequa replied:
Sometimes I see it show `253 / 252` instead of `252 / 252` so this makes me wonder if its really seeing the full batches, do you think it would help to do `steps_per_epoch = len(X_train) / batch_size + 1` just to be safe?

godardt replied:
I think yes. From my part when I'm not sure about that I always refer to the doc or try to find a response from an official repo. I just ran into this: https://github.com/fchollet/keras/pull/6301/files . And as you can see they use `/` instead of `//` on the Keras repo. I also had this issue myself when creating the batching PR, the last bits of the training set was "forgotten" while using `//` (which makes sense). 

I can't tell why but when I trained the model with `/` I've got better results than with `//`. Maybe because of the randomness of events? Maybe because we were dismissing some `primary` and `clear` images in the process? I can't tell.

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -190,119 +191,121 @@
+x = Flatten()(x)
+output = Dense(17, activation='sigmoid')(x)
+
+model = Model(input_tensor, output)
```
We could move all this (line 218 to 227) to its own file as you did with `densenet169.py`. We can name it `vgg16.py` perhaps? :)

EKami commented:
https://github.com/EKami/planet-amazon-deforestation/pull/14#discussion_r128934730

@jamesrequa I also tried to use LROnPlateau myself and it worked pretty well (got approximatively the same accuracy as the original code when running it on 20 epochs) but it's incompatible with `EarlyStopping` be careful. As the latter will stop the training if it sees that it plateau and ReduceLROnPlateau will try to reduce the learning rate.

jamesrequa committed:
removed unneeded files

jamesrequa committed:
remove commented out line

jamesrequa committed:
added weights folder

EKami commented:
You added `notebooks/weights/weights.best.hdf5` . This should be added to `.gitignore` (so you don't commit large files (this one is 30mb)). You should only add a junk file to the `notebooks/weights/` folder such as `notebooks/weights/.DS_Store` :)

jamesrequa committed:
create vgg16.py remove keras_helper

Created vgg16.py file with following functions:
- `create_model`: create pre-trained VGG16 model with new top layers
added
- `predict`: generate trained model predictions on test dataset
- `map_predictions`: return predictions mapped to labels

Removed keras_helper file since it is no longer needed

jamesrequa committed:
Update .gitignore

Updated weights file so that it is now contained within weights folder

jamesrequa committed:
remove previous weight file

jamesrequa committed:
Update .gitignore

jamesrequa committed:
.keep

jamesrequa committed:
removing tensorflow.contrib.keras.api

jamesrequa committed:
notebook updates

- moved model, predict, map_predictions, and fbeta_score all to
separate file `vgg16.py`
- remove keras_helper import
- bringing back history callback and plot curve
- Removed `for loop` learning rate scheduler
- Adding `min_lr` to `ReduceLROnPlateau` to make it more compatible
with `EarlyStopping`
- Changed `factor` to 0.1 for `ReduceLROnPlateau` so learning rate
reduces 10x on each update
- `steps_on_epochs` corrected

jamesrequa commented:
@EKami All of the requested changes for this PR have now been completed. Let me know if all looks good for merging :)

godardt commented on .gitignore:
```diff
@@ -2,7 +2,8 @@
 notebooks/amazon_forest_notebook.ipynb
-notebooks/weights.best.hdf5
+notebooks/weights/*
+!notebooks/weights/.keep
```
Oh neat! I didn't know about this `!` 😄

jamesrequa replied:
Yep this makes it so any file in weights folder will not get loaded

EKami commented on src/keras_helper.py:
```diff
@@ -1,157 +0,0 @@
-import numpy as np
```
Bye bye little one 😢

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -190,110 +191,78 @@
-batch_size = 64
-epochs_arr = [35, 15, 5]
-learn_rates = [0.002, 0.0002, 0.00002]
+model = vgg16.create_model(img_dim=(128, 128, 3))
```
This could be moved to `vgg16.py`, it's vgg related. Then you could create a method `get_summary()` to print the summary below :)

jamesrequa replied:
We are still currently using keras in the notebook for the model functions like `compile` and `fit_generator`, so unless we moved literally everything in vgg16.py I think its OK to just leave it as is. Do you think we should move everything to the helper file? This makes everything a lot more high level in the notebook but also hides a lot of the inner details of the network functions (making it more challenging for the user to tune the network within the notebook).

godardt replied:
Well ideally the notebook should only be used to present results, not to tune anything but I didn't respect this rule from the very start so I don't know. Put (or leave) this line where you think is best :)

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -190,110 +191,78 @@
+             EarlyStopping(monitor='val_loss', patience=3, verbose=1, min_delta=1e-4),
+             ReduceLROnPlateau(monitor='val_loss', factor=0.1, patience=1, cooldown=0, min_lr=1e-7, verbose=1),
+             ModelCheckpoint(filepath='weights/weights.best.hdf5', verbose=1, save_best_only=True, 
+                             save_weights_only=True, mode='auto')]
```
We can also move this to `vgg16.py`. Not sure tho. Your call :)

godardt commented on .gitignore:
```diff
@@ -2,7 +2,8 @@
 input/*
 submission_file.csv
 notebooks/amazon_forest_notebook.ipynb
-notebooks/weights.best.hdf5
```
You're still missing this one now. We need to put it back 😛

godardt commented on notebooks/amazon_forest_notebook.py:
```diff
@@ -190,110 +191,78 @@
+plt.ylabel('loss')
+plt.xlabel('epoch')
+plt.legend(['train', 'validation'], loc='upper left')
+plt.show()
```
Very neat! Thanks :D

godardt reviewed (changes requested):
This PR is awesome! Just 2-3 little things to change and we're good 🎉

EKami commented:
<img width="564" alt="amazon_forest_notebook" src="https://user-images.githubusercontent.com/4030626/28619849-b800f668-720a-11e7-8a06-30886faa04d8.png">
There is something not right here. Is it?

EKami commented:
<img width="993" alt="amazon_forest_notebook" src="https://user-images.githubusercontent.com/4030626/28619893-eb4d960c-720a-11e7-870b-3f206de0da0c.png">
Btw the training stopped earlier than expected just after the `ReduceLROnPlateau` has decreased the learning rate. But I believe it's expected?

jamesrequa commented:
@EKami 
For your first question, on the plot yes this is to be expected as the network quickly drops to a low loss and gets less and less decrease in loss over subsequent epochs. If we were to stop training at "convergence" which is much earlier in the epoch cycle we would end up with a much lower fbeta score. So in this case it makes sense to let it continue training according to our schedule set by `EarlyStopping` and`ReduceLROnPlateau`.

For your second question, yes the training has stopped where it should be after `ReduceLROnPlateau` has decreased to its minumum learning rate value (`min_lr=1e-7`).

EKami commented:
@jamesrequa Oh I see, thanks for these explanations :)
